packages/jest-diff/src/diff_strings.js

```diff
@@ -34,45 +36,120 @@ type Hunk = {|
   oldStart: number,
 |};
 
-const getColor = (added: boolean, removed: boolean) =>
-  added ? chalk.red : removed ? chalk.green : chalk.dim;
+type DIFF_D = -1 | 1 | 0; // diff digit: removed | added | equal
+
+// Given chunk, return diff character.
+const getC = (chunk): string => (chunk.removed ? '-' : chunk.added ? '+' : ' ');
+
+// Given diff character by getC from chunk or line from hunk, return diff digit.
+const getD = (c: string): DIFF_D => (c === '-' ? -1 : c === '+' ? 1 : 0);
 
-const getBgColor = (added: boolean, removed: boolean) =>
-  added ? chalk.bgRed : removed ? chalk.bgGreen : chalk.dim;
+// Color for text of line.
+// If compared lines are equal and expected and received are data structures,
+// then delta is difference in length of original lines.
+const getColor = (d: DIFF_D, delta?: number) =>
+  d === 1 ? chalk.red : d === -1 ? chalk.green : delta ? chalk.cyan : chalk.dim;
 
+// Do NOT color leading or trailing spaces if original lines are equal:
+
+// Background color for leading or trailing spaces.
+const getBgColor = (d: DIFF_D) =>
+  d === 1 ? chalk.bgRed : d === -1 ? chalk.bgGreen : chalk.bgCyan;
+
+// ONLY trailing if expected is snapshot or multiline string.
 const highlightTrailingWhitespace = (line: string, bgColor: Function): string =>
   line.replace(/\s+$/, bgColor('$&'));
 
+// BOTH leading AND trailing if expected and received are data structures.
+const highlightWhitespace = (line: string, bgColor: Function): string =>
+  highlightTrailingWhitespace(line.replace(/^\s+/, bgColor('$&')), bgColor);
+
 const getAnnotation = (options: ?DiffOptions): string =>
   chalk.green('- ' + ((options && options.aAnnotation) || 'Expected')) +
   '\n' +
   chalk.red('+ ' + ((options && options.bAnnotation) || 'Received')) +
   '\n\n';
 
-const diffLines = (a: string, b: string): Diff => {
+// Given string, return array of its lines.
+const splitIntoLines = string => {
+  const lines = string.split('\n');
+
+  if (lines.length !== 0 && lines[lines.length - 1] === '') {
+    lines.pop();
+  }
+
+  return lines;
+};
+
+// Given diff character and corresponding compared line, return line with colors
+// and original indentation if compared with `indent: 0` option.
+const formatLine = (
+  c: string,
+  lineCompared: string,
+  getOriginal?: Function,
+) => {
+  const d = getD(c);
+
+  if (getOriginal) {
+    // getOriginal callback function if expected and received are data structures.
+    const gotOriginal = getOriginal(d);
+    const lineOriginal = d === 0 ? gotOriginal[1] : gotOriginal;
+    const lengthOriginal = lineOriginal.length;
+    // If compared lines are equal,
+    // then delta is difference in length of original lines.
+    const delta = d === 0 ? lengthOriginal - gotOriginal[0].length : 0;
+    return getColor(d, delta)(
```

Comment:
This function call is also incredibly hard to read. Any way we could factor out the inline ternaries etc.?